### PR TITLE
create recv bind config

### PR DESCRIPTION
**Pull request by stefan-langenmaier**
I'm running a Galera cluster behind a NAT and the machine has no device with the public IP address. So when an IST should run there is the problem described in issue #105 and I get the following error:

```
    WSREP: Failed to prepare for incremental state transfer: Failed to open IST listener at ssl://example.org:4568', asio error 'Cannot assign requested address': 99 (Cannot assign requested address)
     at galera/src/ist.cpp:prepare():356. IST will be unavailable.
```

This pull request adds a configuration option "ist.recv_bind" that defaults to "ist.recv_addr" but when it is set it specifies the address Galera will bind and listen on for IST. The "ist.recv_addr" will be unaffected and the donor will receive this address as before.

The unit tests pass, it works on my setup, but I'm no C++ developer so I'm open to make modifications. :)

### CODE REVIEW

**Comment by klausenbusk** [comment]
Ping @temeo .. What is the status? It costly to do SST all the time.. At least the data-set is only 1,2GB

File: galera/src/ist.cpp [hunk 1]
```diff
@@ -204,9 +216,98 @@ IST_determine_recv_addr (gu::Config& conf)
...
+		else
+			recv_bind.insert(0, "tcp://");
+	}
+
+	gu::URI rb_uri(recv_bind);
+
+	// Cargo-culting
+	// no idea what I'm doing
+	if (!conf.has(galera::BASE_HOST_KEY))
+		conf.set(galera::BASE_HOST_KEY, rb_uri.get_host());
```
Comment (temeo) [inline galera/src/ist.cpp]: This would override base_host config parameter so it is not right here.

Comment (stefan-langenmaier) [inline galera/src/ist.cpp]: Thanks for the comment. What should be the strategy? Should I remove the "if"-block?

**Comment by stefan-langenmaier** [comment]
updated the pull request according to the comment

File: galera/src/ist.cpp [hunk 2]
```diff
@@ -204,9 +216,102 @@ IST_determine_recv_addr (gu::Config& conf)
...
 }
 
+static std::string IST_determine_recv_bind(gu::Config& conf) {
+	std::string recv_bind;
+
+	try
+	{
+		recv_bind = conf.get(galera::ist::Receiver::RECV_BIND);
+	}
+	catch (gu::NotFound&)
```
Comment (ayurchen) [inline galera/src/ist.cpp]: I think that since you add this parameter to config set above, here you must be catching gu::NotFound is pointless.
Also, instead of duplicating the whole process of getting recv_addr from config, you could just pass already determined recv_addr to this function, or see below

File: galera/src/ist.cpp [hunk 3]
```diff
@@ -215,10 +320,12 @@ galera::ist::Receiver::prepare(wsrep_seqno_t first_seqno,
     ready_ = false;
     version_ = version;
     recv_addr_ = IST_determine_recv_addr(conf_);
-    gu::URI     const uri(recv_addr_);
+    recv_bind_ = IST_determine_recv_bind(conf_);
```
Comment (ayurchen) [inline galera/src/ist.cpp]: or you could simply

```
try
{
    recv_bind_ = IST_determine_recv_bind(conf_);
}
catch (gu::NotSet&)
{
    resv_bind_ = recv_addr_;
}
```

and this way remove any concern about recv_addr from `IST_determine_recv_bind()`

File: scripts/packages/README [hunk 4]
```diff
@@ -443,6 +443,12 @@ recv_addr
     turns on incremental state transfer. IST will use SSL if SSL is configured
     as described above. No default.
 
+recv_bind
```
Comment (ayurchen) [inline scripts/packages/README:446]: And I'd call the parameter `ist_bind`. So that there is no ambiguity.

File: galera/src/ist.cpp (hunk 2 above)
Comment (stefan-langenmaier) [inline galera/src/ist.cpp]: I'm all for removing unused code. The catch for gu::NotFound should no longer be there.

File: galera/src/ist.cpp (hunk 3 above)
Comment (stefan-langenmaier) [inline galera/src/ist.cpp]: I prefer this solution. This is pushed.

File: scripts/packages/README (hunk 4 above)
Comment (stefan-langenmaier) [inline scripts/packages/README:446]: I have no problem changing the name of the parameter, I just want to state that I prefer the current naming. It shows the two variables are related as both seem to be used for the IST. And if the name is changed some other variables like uri_addr and uri_bind should also be changed.
What would you propose here?

File: galera/src/ist.cpp [hunk 5]
```diff
@@ -204,9 +216,60 @@ IST_determine_recv_addr (gu::Config& conf)
...
+
+		port += 1;
+
+		recv_bind += ":" + gu::to_string(port);
+	}
+
+	log_info<< "IST receiver bind using " << recv_bind;
+	return recv_bind;
+}
+
```
Comment (temeo) [inline galera/src/ist.cpp:272]: Syntactic issue: This code block seems to use tabs for indentation instead of 4 spaces what the rest of Galera code uses.

Comment (stefan-langenmaier) [inline galera/src/ist.cpp:272]: Should be fixed.

File: scripts/packages/README (hunk 4 above)
Comment (ayurchen) [inline scripts/packages/README:446]: I retract my comment previous here. Of course recv_bind is the proper name. Sorry.

**Comment by stefan-langenmaier** [comment]
Trying to push this again. :-)
Is there anything else I can do? Do you have other comments?

**Comment by radek-senfeld** [comment]
+1

I'm using @stefan-langenmaier patch to solve my IST behind NAT problem (yep, Docker involved)..

Please, accept this patch in an official release..

Thank you very much for doing the hard work, Stefan! You saved my day.. ;-)

**Comment by ayurchen** [comment]
Hi Stefan, I think your patch is good to go. However it is too big to be taken as is (lawyers). So there will have to be one last step. With 3.15 release we added two files:
CONTRIBUTOR_AGREEMENT.txt and CONTRIBUTORS.txt. Putting your name/github account in CONTRIBUTORS.txt **as part of the pull request** will indicate acceptance of Contributor Agreement.
Sorry for the bureaucracy, but lawyers... 

So in short what you need to do is to sign the CONTRIBUTORS.txt, rebase the whole thing into single commit (as all these intermediate steps will only unnecessarily clutter the history) and push the branch once again.

**Comment by stefan-langenmaier** [comment]
Hi,

I rebased it on the 3.x branch, added my name to the CONTRIBUTORS.txt and squashed the commits into a single one. Let me know if there is something else to do.

Best regards
Stefan

**Comment by philip-galera** [comment]
@stefan-langenmaier, thank you very much for your effort. We will document and test this option in view of adding it to our next release.

**Comment by nirbhayc** [comment]
@philip-galera While I do see that this patch was not part of 25.3.15. I guess it will make into 25.3.16. Right?

**Comment by philip-galera** [comment]
Yes, that is correct.
